subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/LatestIntegrationTest.groovy

```diff
@@ -0,0 +1,220 @@
+        mavenRepo.module("org", "foo", '1.1').publish()
+        mavenRepo.module("org", "foo", '1.1-SNAPSHOT').publish()
+        mavenRepo.module("org", "foo", '2.0-SNAPSHOT').publish()
+
+        buildFile << """
+            dependencies {
+                conf 'org:foo:latest.release'
+            }
+        """
+
+        when:
+        run 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":testLatest:") {
+                edge("org:foo:latest.release","org:foo:1.1")
+            }
+        }
+    }
+
+    void 'latest.release and range constraint excluding highest still selects it'() {
+        given:
+        mavenRepo.module("org", "foo", '1.0').publish()
+        mavenRepo.module("org", "foo", '1.1').publish()
+        mavenRepo.module("org", "foo", '1.2').publish()
+        mavenRepo.module("org", "foo", '2.0').publish()
+
+        buildFile << """
+            dependencies {
+                constraints {
+                    conf 'org:foo:[1.0,2.0)'
+                }
+                conf 'org:foo:latest.release'
+            }
+        """
+
+        when:
+        run 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":testLatest:") {
+                edge("org:foo:latest.release","org:foo:2.0")
+                edge("org:foo:[1.0,2.0)","org:foo:2.0")
+            }
+        }
+    }
+
+    void 'latest.release and prefer constraint selects latest'() {
+        given:
+        mavenRepo.module("org", "foo", '1.0').publish()
+        mavenRepo.module("org", "foo", '1.1').publish()
+        mavenRepo.module("org", "foo", '1.2').publish()
+        mavenRepo.module("org", "foo", '2.0').publish()
+
+        buildFile << """
+            dependencies {
+                constraints {
+                    conf('org:foo') {
+                        version {
+                            prefer '1.2'
+                        }
+                    }
+                }
+                conf 'org:foo:latest.release'
+            }
+        """
+
+        when:
+        run 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":testLatest:") {
+                edge("org:foo:latest.release","org:foo:2.0")
+                edge("org:foo:1.2","org:foo:2.0")
+            }
+        }
+    }
+
+    @Ignore('This ends up selecting a snapshot while we have a latest.release')
+    void 'latest.release and prefer constraint with wrong status selects latest'() {
+        given:
+        mavenRepo.module("org", "foo", '1.0').publish()
+        mavenRepo.module("org", "foo", '1.1').publish()
+        mavenRepo.module("org", "foo", '2.0').publish()
+        mavenRepo.module("org", "foo", '2.1-SNAPSHOT').withNonUniqueSnapshots().publish()
+
+        buildFile << """
+            dependencies {
+                constraints {
+                    conf('org:foo') {
+                        version {
+                            prefer '2.1-SNAPSHOT'
```

Comment:
Since we are likely to change the meaning of `prefer`, this test might be better off expressing the dependency as `org:foo:2.1-SNAPSHOT` (which is stronger than `prefer` since it will continue to reject lower versions).

Reply:
That's an interesting comment. I am expecting this test to finally pass when `prefer` no longer has the implicit reject.
However I should indeed add a test where the constraint is expressed as `org:foo:2.1-SNAPSHOT` which should always fail once the bug is fixed.
Makes sense?

Reply:
Yep, with my changes to `prefer` this test should result in `2.0` being selected (so the test as written will not pass, but the build will succeed and choose the correct version).

With `org:foo:2.1-SNAPSHOT`, the test should always fail. We will be effectively making the contract of `latest.<status>` stronger, implying "reject all lower status". I believe this will already work if a status attribute is applied to the dependency.

We'll need to consider how to conflict resolve `org:foo:latest.snapshot` with `org:foo:latest.release`. I guess we'd choose a `release` version, since it would satisfy both. Not sure this is what users would really want, though.